syntax = "proto3";

package cfs;

service ConfigurationServer
{
  rpc GetKeys(Empty) returns (stream Key) {}
  rpc GetValue(Key) returns (Value) {}
  rpc GetDisplayHint(Key) returns (ValueDisplayHint) {}
  rpc SetValue(Pair) returns (Result) {}
}

message Key
{
  string name = 1;
}

message Value
{
  StatusCode status = 1;
  string value = 2;
}

message Pair
{
  string name = 1;
  string value = 2;
}

message IntValueDisplayHint
{
  int32 min = 1;
  int32 max = 2;
}

message RealValueDisplayHint
{
  double min = 1;
  double max = 2;
}

message BoolValueDisplayHint
{
}

message StringValueDisplayHint
{
  repeated string values = 1;
  bool mustBeOneOf = 2;
}

message ValueDisplayHint
{
  StatusCode status = 1;
  bool runtimeMutable = 2;
  string description = 3;
  oneof hint
  {
    IntValueDisplayHint intDisplayHint = 4;
    RealValueDisplayHint realDisplayHint = 5;
    BoolValueDisplayHint boolDisplayHint = 6;
    StringValueDisplayHint stringDisplayHint = 7;
  }
}

message Empty
{
}

enum StatusCode
{
  SUCCESS = 0;
  KEYNOTFOUND = 1;
  VALUEINVALID = 2;
}

message Result
{
  StatusCode status = 1;
}
